### PR TITLE
k-Eigenvalue Solver Transplant

### PR DESCRIPTION
**Pull request by zachhardy**
## Notes

This PR is a transplant to incorporate the k-eigenvalue solver found within the `LBKEigenvalueSolver` repository of the chi-tech organization. This is being done due to `KEigenvalueSolver` becoming obsolete after many rounds of changes to `ChiTech`. This solver has recently only changed to stay up-to-date with `ChiTech`, and not to add additional features for the solver itself. Housing this here will allow for simpler maintenance. This PR is based off of #203 and should be addressed once that is merged.

### CODE REVIEW

File: ChiModules/LBKEigenvalueSolver/lua/lbkes_execute.cc [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  int solver_index = lua_tonumber(L, 1);
+
+  //============================================= Get pointer to solver
+  chi_physics::Solver* psolver;
+  KEigenvalueSolver* solver;
+  try
+  {
+    psolver = chi_physics_handler.solver_stack.at(solver_index);
+
+    solver = dynamic_cast<KEigenvalueSolver*>(psolver);
```
Comment (Naktakala) [inline ChiModules/LBKEigenvalueSolver/lua/lbkes_execute.cc]: Please make utility routines for this similar to what LBS had.

File: ChiModules/LBKEigenvalueSolver/lua/lbkes_initialize.cc [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  int solver_index = lua_tonumber(L, 1);
+
+  //================================================== Get pointer to solver
+  chi_physics::Solver* psolver;
+  KEigenvalueSolver* solver;
+  try
+  {
+    psolver = chi_physics_handler.solver_stack.at(solver_index);
+
+    solver = dynamic_cast<KEigenvalueSolver*>(psolver);
```
Comment (Naktakala) [inline ChiModules/LBKEigenvalueSolver/lua/lbkes_initialize.cc]: Ditto

**Review by Naktakala — CHANGES_REQUESTED** [review]
Just the one change please.

**Comment by zachhardy** [comment]
Done @Naktakala

File: ChiModules/LBKEigenvalueSolver/lua/lbkes_execute.cc [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+//###################################################################
+/**Executes the LBS solver.
+\param SolverIndex int Handle to the solver.
+ \ingroup LuaLBS
+ */
+int chiLBKESExecute(lua_State *L)
+{
+  //============================================= Get pointer to solver
+  int solver_index = lua_tonumber(L,1);
+  auto lbkes_solver = LinearBoltzmann::lua_utils::
```
Comment (Naktakala) [inline ChiModules/LBKEigenvalueSolver/lua/lbkes_execute.cc]: Mmmmm. This util, `LinearBoltzmann::lua_utils::GetSolverByHandle `, is ambiguous if the `lbs_lua_utils.h` happens to be in the same header chain. Some options to resolve this:
- Segregate the utils by changing the KEigen-specific namespace from `lua_utils` to `k_eigen_lua_utils` (or something like it)
- Change the util's name from `GetSolverByHandle` to `GetKEigenvalueSolverByHandle`

I am tilting slight toward the first option cause that will allow you room to expand KEigen-specific utils in future.

File: ChiModules/LBKEigenvalueSolver/lua/lbkes_setproperty.cc [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  LuaCheckNilValue(__FUNCTION__, L, 1);
+  int solver_index = lua_tonumber(L, 1);
+
+  //============================================= Get pointer to solver
+  chi_physics::Solver* psolver;
+  KEigenvalueSolver* solver;
+  try
+  {
+    psolver = chi_physics_handler.solver_stack.at(solver_index);
+    solver = dynamic_cast<KEigenvalueSolver*>(psolver);
```
Comment (Naktakala) [inline ChiModules/LBKEigenvalueSolver/lua/lbkes_setproperty.cc]: No love? You just plumbed in a new bathroom for your bride ... make her take a shower in it.

Aka. Use the utils. :-)

**Review by Naktakala — CHANGES_REQUESTED** [review]

**Comment by zachhardy** [comment]
Agreed with the renaming of the namespace and oops on the missing the old solver getter in `setproperty`. Should be good now.

**Review by Naktakala — APPROVED** [review]

**Pull request merged**